File: 2-add-two-numbers/add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        sum = 0
        r = p = ListNode(0)
        while l1 or l2:
            if l1:
                sum += l1.val
                l1 = l1.next
            if l2:
                sum += l2.val
                l2 = l2.next
            p.next = ListNode(sum%10)
            sum = int(sum/10)
            p = p.next
            
            if sum:
                p.next = ListNode(sum)
        # 此写法也可
        # if sum:
        #     p.next = ListNode(sum)
            
        return r.next

File: 2-add-two-numbers/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int sum=0;
        ListNode p = new ListNode(0);
        ListNode r = p;
        
        while(l1!=null || l2!=null){
            if(l1!=null){
                sum+=l1.val;
                l1 = l1.next;
            }
            if(l2!=null){
                sum+=l2.val;
                l2 = l2.next;
            }
            p.next = new ListNode(sum %10);
            sum = sum /10;
            p = p.next;
        }
        if(sum==1){
             p.next = new ListNode(sum);
        }
        return r.next;
        
    }
}